testing/shared/e2e/performance/benchmarking.ts

```diff
@@ -0,0 +1,334 @@
+
+import type { BrowserCommunications } from '../../playwright.utils';
+import { gotoUrl, waitFor } from '../../playwright.utils';
+
+export type Framework = 'typescript' | 'reactFunctionalTs';
+export type CustomVersion = `v${number}.${number}.${number}`;
+export type Version = 'prod' | 'staging' | 'local' | CustomVersion;
+export type Entry<T> = T extends readonly (infer U)[] ? U : T extends object ? T[keyof T] : T;
+
+/**
+ * Describes a performance benchmarking test suite.
+ */
+export type Describe = {
+    minIterations?: number; // default is 10, also used as an inner loop iteration count
+    maxIterations?: number; // default is 1000
+    testCases: TestCase[];
+    timeout?: number; // in milliseconds, default is 3 minutes
+};
+
+/**
+ * Describes a single test case within a performance benchmarking suite.
+ */
+export type TestCase = {
+    name: string;
+    /** @deprecated don't forget to re-enable your test */
+    skip?: boolean;
+    framework: Framework;
+    control: Variant;
+    variant: Variant;
+    preSetup?: (page: Page) => Promise<void>;
+    setupPreActions?: (page: Page) => Promise<void>;
+    actions?: (page: Page) => Promise<void>;
+    expectsPostActions?: (page: Page, comms: BrowserCommunications) => Promise<void>;
+    metrics?: Entry<(typeof PerformanceObserver)['supportedEntryTypes']>;
+};
+
+/**
+ * Describes a variant of a test case, which can include a specific URL and version.
+ *
+ * @prop url can be a plunker, or any grid example url.
+ * @prop version is used to determine the version of the grid to test against, e.g. 'prod', 'staging', 'local', or a specific version like 'v29.0.0'.
+ *               Playwright will intercept whatever version is specified in the example, and use this version instead.
+ * @prop cookies are used to set cookies for the test case, e.g. for plunker acceptance cookie
+ */
+export type Variant = {
+    url?: string;
+    version: Version;
+    shouldInjectScript?: boolean;
+    cookies?: Parameters<BrowserContext['addCookies']>[0];
+};
+
+dotenv.config({ path: path.join(__dirname, '../../../../documentation/ag-grid-docs/.env.dev') }); // grab docs PORT
+
+const knownUrls: Record<Version, string> = {
+    local: `https://localhost:${process.env.PORT || '4610'}`,
+    staging: 'https://grid-staging.ag-grid.com',
+    prod: 'https://www.ag-grid.com',
+};
+
+export const agChartsVersion = JSON.parse(
+    fs.readFileSync(path.join(__dirname, '../../../../packages/ag-grid-enterprise/package.json')).toString()
+).optionalDependencies['ag-charts-enterprise'];
+
+/**
+ * Taken from ag-grid-enterprise package.json git history
+ */
+const gridToChartsMap = {
+    local: agChartsVersion,
+    prod: 'v11.3.0',
+    staging: agChartsVersion,
+    'v33.3.0': 'v11.3.0',
+    'v33.2.3': 'v11.2.3',
+    'v33.2.1': 'v11.2.1',
+    'v33.1.1': 'v11.1.1',
+    'v33.1.0': 'v11.1.0',
+    'v33.0.1': 'v11.0.0',
+    'v32.2.0': 'v11.0.0', // was latest, changed to the closest one
```

Comment:
I would have thought this should be the last of v10

Reply:
```suggestion
    'v32.2.0': 'v10.1.0', // was latest, changed to the closest one
```
I didn't test it, but makes sense